File: common/src/main/java/juuxel/adorn/item/AdornBookItem.java
package juuxel.adorn.item;

import juuxel.adorn.networking.OpenBookS2CMessage;
import juuxel.adorn.platform.PlatformBridges;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.Identifier;
import net.minecraft.stats.Stats;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.item.component.TooltipDisplay;
import net.minecraft.world.level.Level;

import java.util.function.Consumer;

public final class AdornBookItem extends Item {
    public static BookTooltipProvider tooltipProvider = (_, _) -> {};

    private final Identifier bookId;

    public AdornBookItem(Identifier bookId, Properties settings) {
        super(settings);
        this.bookId = bookId;
    }

    @Override
    public InteractionResult use(Level world, Player user, InteractionHand hand) {
        if (!world.isClientSide()) {
            PlatformBridges.get().getNetwork().sendToClient(user, new OpenBookS2CMessage(bookId));
        }
        user.awardStat(Stats.ITEM_USED.get(this));

        return InteractionResult.SUCCESS;
    }

    @Override
    public void appendHoverText(ItemStack stack, TooltipContext context, TooltipDisplay displayComponent, Consumer<Component> textConsumer, TooltipFlag type) {
        super.appendHoverText(stack, context, displayComponent, textConsumer, type);
        tooltipProvider.appendTooltip(bookId, textConsumer);
    }

    @FunctionalInterface
    public interface BookTooltipProvider {
        void appendTooltip(Identifier bookId, Consumer<Component> builder);
    }
}

File: common/src/main/kotlin/juuxel/adorn/item/AdornBookItem.kt
package juuxel.adorn.item

import juuxel.adorn.platform.PlatformBridges
import net.minecraft.client.item.TooltipContext
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemGroup
import net.minecraft.item.ItemStack
import net.minecraft.stat.Stats
import net.minecraft.text.Text
import net.minecraft.util.ActionResult
import net.minecraft.util.Formatting
import net.minecraft.util.Hand
import net.minecraft.util.Identifier
import net.minecraft.util.TypedActionResult
import net.minecraft.world.World

class AdornBookItem(private val bookId: Identifier, settings: Settings) : Item(settings) {
    override fun use(world: World, player: PlayerEntity, hand: Hand): TypedActionResult<ItemStack> {
        if (!world.isClient) {
            PlatformBridges.network.sendOpenBookPacket(player, bookId)
        }
        player.incrementStat(Stats.USED.getOrCreateStat(this))

        return TypedActionResult(ActionResult.SUCCESS, player.getStackInHand(hand))
    }

    override fun appendTooltip(stack: ItemStack, world: World?, texts: MutableList<Text>, context: TooltipContext) {
        super.appendTooltip(stack, world, texts, context)
        val bookManager = PlatformBridges.resources.bookManager
        if (bookId in bookManager) {
            texts.add(Text.translatable("book.byAuthor", bookManager[bookId].author).formatted(Formatting.GRAY))
        }
    }

    override fun isIn(group: ItemGroup?) = AdornItems.isIn(group, this)
}
